src/alchemlyb/workflows/abfe.py

```diff
@@ -0,0 +1,684 @@
+    T : float
+        Temperature in K. Default: 298.
+    outdirectory : str
+        Directory in which the output files produced by this script will be
+        stored. Default: os.path.curdir.
+    ignore_warnings : bool
+        Turn all errors into warnings.
+
+    Attributes
+    ----------
+    logger : Logger
+        The logging object.
+    file_list : list
+        The list of filenames sorted by the lambda state.
+    '''
+    def __init__(self, units='kT', software='Gromacs', dir=os.path.curdir,
+                 prefix='dhdl', suffix='xvg', T=298,
+                 outdirectory=os.path.curdir,
+                 ignore_warnings=False):
+
+        super().__init__(units, software, T, outdirectory)
+        self.ignore_warnings = ignore_warnings
+        self.logger = logging.getLogger('alchemlyb.workflows.ABFE')
+        self.logger.info('Initialise Alchemlyb ABFE Workflow')
+        self.logger.info(f'Alchemlyb Version: f{__version__}')
+        self.logger.info(f'Set Temperature to {T} K.')
+        self.logger.info(f'Set Software to {software}.')
+
+        self.update_units(units)
+
+        self.logger.info(f'Finding files with prefix: {prefix}, suffix: '
+                         f'{suffix} under directory {dir} produced by '
+                         f'{software}')
+        self.file_list = glob(dir + '/**/' + prefix + '*' + suffix,
+                         recursive=True)
+
+        self.logger.info(f'Found {len(self.file_list)} xvg files.')
+        self.logger.info("Unsorted file list: \n %s", '\n'.join(
+            self.file_list))
+
+        if software.lower() == 'gromacs':
+            self.logger.info(f'Using {software} parser to read the data.')
+            self._extract_u_nk = gmx.extract_u_nk
+            self._extract_dHdl = gmx.extract_dHdl
+        elif software.lower() == 'amber':
+            self._extract_u_nk = amber.extract_u_nk
+            self._extract_dHdl = amber.extract_dHdl
+        else: # pragma: no cover
+            raise NotImplementedError(f'{software} parser not found.')
+
+    def read(self):
+        '''Read the u_nk and dHdL data from the
+        :attr:`~alchemlyb.workflows.ABFE.file_list`
+
+        Attributes
+        ----------
+        u_nk_list : list
+            A list of :class:`pandas.DataFrame` of u_nk.
+        dHdl_list : list
+            A list of :class:`pandas.DataFrame` of dHdl.
+        '''
+        u_nk_list = []
+        dHdl_list = []
+        for file in self.file_list:
+            try:
+                u_nk = self._extract_u_nk(file, T=self.T)
+                self.logger.info(
+                    f'Reading {len(u_nk)} lines of u_nk from {file}')
+                u_nk_list.append(u_nk)
+            except: # pragma: no cover
+                msg = f'Error reading read u_nk from {file}.'
+                if self.ignore_warnings:
+                    self.logger.warning(msg)
+                else:
+                    raise ValueError(msg)
```

Comment:
please address

Reply:
To be honest, as soon as I see "pragma: no cover" I know that eventually I'll see this pop up as not working as thought... can you at least manually trigger it and confirm that it does want it should be doing?

And it is important to keep the original exception.

Reply:
I have added the test to test this. The original exception is recorded in

```
except Exception as exc: 
    msg = f'Error reading read u_nk from {file}.\n{exc}'
    raise ValueError(msg)
```

where the {exc} in msg is the original exception.

Reply:
Testing is fantastic!

I still don't like raising `ValueError` here. I'd rather see what really happened _and_ also have access to the actual exception. I also don't see it as a problem with values — it's more likely related to I/O. I just don't see a compelling reason to force it to be a ValueError.

Reply:
Sorry, I cannot figure out how to reraise the same exception. I think one thing that I could do is to use exception Chaining.

So I do

```
            except Exception as exc:
                msg = f'Error reading read u_nk from {file}.'
                if self.ignore_warnings:
                    self.logger.exception(msg +
                        'This exception is being ignored because ignore_warnings=True.')
                else:
                    self.logger.error(msg)
                    raise ValueError(msg) from exc
```
 Note the `raise ValueError(msg) from exc` I will get
```
Traceback (most recent call last):
File "<input>", line 1, in <module>
ValueError: Unable to read u_nk.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/Users/zhiyiwu/GitHub/alchemlyb/src/alchemlyb/workflows/abfe.py", line 122, in read
ValueError: Error reading read u_nk from /Users/zhiyiwu/GitHub/alchemtest/src/alchemtest/gmx/benzene/Coulomb/0000/dhdl.xvg.bz2
```

Reply:
I thought just a bare `raise` will simply re-raise the exception. Does the following not work
```python
            except Exception as exc:
                msg = f'Error reading read u_nk from {file}.'
                if self.ignore_warnings:
                    self.logger.exception(msg +
                        'This exception is being ignored because ignore_warnings=True.')
                else:
                    self.logger.error(msg)
                    raise
```

Reply:
Or 
```python
            except Exception as exc:
                msg = f'Error reading read u_nk from {file}.'
                if self.ignore_warnings:
                    self.logger.exception(msg +
                        'This exception is being ignored because ignore_warnings=True.')
                else:
                    self.logger.raise(msg)
                    raise
```
so that the logger includes everything automatically.

Reply:
Thanks for the suggestion. The raise would work but one would lose the error message saying which file gives the problem.
The log file does contain all the information but I think it might be easier to just dump everything into the console so the user doesn't need to open the log file to do the debugging. The exception chaining will preserve the original exception and just append an additional exception at the end which tells the user which file contains issue.

Reply:
I see the concern that we don't have the filename included. 

What I still don't like is that the actual error that is raised in the end is the `ValueError`.  Here is an example where a `TypeError` is the "inner error":

```python
In [2]: def bad():
   ...:     try:
   ...:         raise TypeError("bad type")
   ...:     except Exception as exc:
   ...:         raise ValueError(f"Oh no {exc} happened!") from exc
   ...:

In [3]: bad()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [2], in bad()
      2 try:
----> 3     raise TypeError("bad type")
      4 except Exception as exc:

TypeError: bad type

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 bad()

Input In [2], in bad()
      3     raise TypeError("bad type")
      4 except Exception as exc:
----> 5     raise ValueError(f"Oh no {exc} happened!") from exc

ValueError: Oh no bad type happened!
```
and I cannot catch the TypeError
```python
In [5]: try:
   ...:     bad()
   ...: except TypeError:
   ...:     print("Caught the TypeError!")
   ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [2], in bad()
      2 try:
----> 3     raise TypeError("bad type")
      4 except Exception as exc:

TypeError: bad type

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
Input In [5], in <cell line: 1>()
      1 try:
----> 2     bad()
      3 except TypeError:
      4     print("Caught the TypeError!")

Input In [2], in bad()
      3     raise TypeError("bad type")
      4 except Exception as exc:
----> 5     raise ValueError(f"Oh no {exc} happened!") from exc

ValueError: Oh no bad type happened!
```

There's a hacky way to maintain the original exception type:
```python
In [6]: def bad():
   ...:     try:
   ...:         raise TypeError("bad type")
   ...:     except Exception as exc:
   ...:         raise type(exc)(f"Oh no {exc} happened!") from exc
   ...:
```

Maybe there's a better way by building a new exception from `exc` that has the file information appended. Something like
```python
args = (exc.args[0] + f" --- Error occured in file {filename}."  ,) + exc.args[1:]
new_exc = type(exc)(*args)
raise new_exc
```

Reply:
I see your concern now. Sorry I initially didn't see the reason that the original exception needs to be preserved. I have updated the code to do `raise type(exc)(msg)`

Reply:
`raise type(exc)(msg)` might not be enough information for downstream where exceptions might be processed. I imagine an IOError that comes with an error code — these are part of the exception. Calling code might really want to know if this EACCESS, NOMEM or whatever the codes are. I think by rebuilding the exception as above we can preserve _all_ information.

Sorry for the long discussion.

Reply:
Sorry, the `msg` contains the exception

```
            except Exception as exc:
                msg = f'Error reading read u_nk from {file}.\n{exc}'
                if self.ignore_warnings:
                    self.logger.exception(msg +
                        'This exception is being ignored because ignore_warnings=True.')
                else:
                    self.logger.exception(msg)
                    raise type(exc)(msg)
```

Reply:
The message is good but cannot be used programmatically. For example, more complex exceptions can be raised
```python
>>> raise IOError(errno.ENOENT, "The file does not exist", "filename.txt")
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Input In [39], in <cell line: 1>()
----> 1 raise IOError(errno.ENOENT, "The file does not exist", "filename.txt")

FileNotFoundError: [Errno 2] The file does not exist: 'filename.txt'
```
Note that (1) the errno changes the IOError into a FileNotFoundError and that (2) the exception has additional attributes such as errno and filename
```python
>>> exc = IOError(errno.ENOENT, "The file does not exist", "filename.txt")
>>> print(exc.filename, exc.errno)
filename.txt 2
```
all of which get lost when we just use the textual message. 

With the correct exception raised, calling code can trap any IOError and directly figure which file was responsible. I don't know how many people want to do this but why should we limit the possibilities? Transparent exceptions seem to be a good thing to me.
```

Reply:
Thanks for the comment. I have fiddled with this a bit and I think exception chaining could do this in an elegant fashion.
So instead of doing `raise ValueError(msg) from exc`, where `ValueError` is the final exception, I do `raise exc from ValueError(msg)`, in this case, the original exception `exc` is the final exception and the exception that shows which file gives error will be conserved, which also avoids editing the exception message to combine the two.